sound/soc/sof/ipc4-pcm.c

```diff
@@ -37,72 +64,77 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	struct sof_ipc4_pipeline *pipeline;
-	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
-	int ret = 0;
-	int num_widgets;
+	int ret, i, j;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
-	list = spcm->stream[substream->stream].list;
+	pipeline_info = &spcm->stream[substream->stream].pipeline_info;
+	data.count = 0;
 
-	for_each_dapm_widgets(list, num_widgets, widget) {
-		swidget = widget->dobj.private;
-
-		if (!swidget)
-			continue;
+	/*
+	 * IPC4 requires pipelines to be triggered in order starting at the sink and
+	 * walking all the way to the source. So add the pipeline instance ID's starting with the
+	 * last one in the trigger list. Skip those pipelines that are already in the requested
+	 * state
+	 */
+	for (i = pipeline_info->count - 1; i >= 0; i--) {
+		pipe_widget = pipeline_info->pipelines[i];
+		pipeline = pipe_widget->private;
 
 		/*
-		 * set pipeline state for both FE and BE pipelines for RUNNING state.
-		 * For PAUSE/RESET, set the pipeline state only for the FE pipeline.
+		 * skip triggering pipelines that are already in the requested state. This will
+		 * be true for backend pipelines that are triggered in the BE DAI trigger.
 		 */
-		switch (state) {
-		case SOF_IPC4_PIPE_PAUSED:
-		case SOF_IPC4_PIPE_RESET:
-			if (!WIDGET_IS_AIF(swidget->id))
-				continue;
-			break;
-		default:
-			break;
-		}
+		if (pipeline->state != state)
+			data.pipeline_ids[data.count++] = pipeline_info->pipelines[i]->instance_id;
+	}
 
-		/* find pipeline widget for the pipeline that this widget belongs to */
-		pipeline_widget = swidget->pipe_widget;
-		pipeline = (struct sof_ipc4_pipeline *)pipeline_widget->private;
-
-		if (pipeline->state == state)
-			continue;
-
-		/* first set the pipeline to PAUSED state */
-		if (pipeline->state != SOF_IPC4_PIPE_PAUSED) {
-			ret = sof_ipc4_set_pipeline_state(sdev, pipeline_widget->instance_id,
-							  SOF_IPC4_PIPE_PAUSED);
-			if (ret < 0) {
-				dev_err(sdev->dev, "failed to pause pipeline %d\n",
-					swidget->pipeline_id);
-				return ret;
-			}
-		}
+	/* return if all pipelines are in the requested state already */
+	if (!data.count)
+		return 0;
 
-		pipeline->state = SOF_IPC4_PIPE_PAUSED;
+	/* pause all pipelines */
+	ret = sof_ipc4_set_multi_pipeline_state(sdev, SOF_IPC4_PIPE_PAUSED, &data);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to pause all pipelines\n");
+		return ret;
+	}
 
-		if (pipeline->state == state)
-			continue;
+	/* update PAUSED state for all pipelines that were just triggered */
+	for (i = 0; i < pipeline_info->count; i++) {
+		pipe_widget = pipeline_info->pipelines[i];
+		pipeline = pipe_widget->private;
 
-		/* then set the final state */
-		ret = sof_ipc4_set_pipeline_state(sdev, pipeline_widget->instance_id, state);
-		if (ret < 0) {
-			dev_err(sdev->dev, "failed to set state %d for pipeline %d\n",
-				state, swidget->pipeline_id);
-			break;
+		for (j = 0; j < data.count; j++) {
```

Comment:
Based on the comment on front of the topmost for() loop, it would be logical to use the `data.count` as the first loop (the pipelines that got triggered) and the inner loop to use `pipeline_info->count`

Reply:
Or just reword the comment, so we will not need to reach down to `pipeline` in every inner loop?

Reply:
@ujfalusi can this be a fixup? I dont want to lose Rander/Bard's approval